Define support requirements for integers (#749)
httpwg/http-extensions: draft-ietf-httpbis-header-structure.md

Example-IntegerHeader: 42
~~~

Parsers MUST accurately support integers in the range -9,007,199,254,740,991 to
9,007,199,254,740,991 inclusive (i.e., -2^53+1 to 2^53-1) at a minimum. Parsers MUST fail parsing

> on values not within the range supported accurately.

Reviewer: This reads a little awkward to me. Perhaps something like "Implementations must have a range for supported integers. At a minimum this range must be X to Y. Parsers must fail on parsing values outside the range for supported integers."

> values outside this range if they are not supported by the parser.

Reviewer 2: I think that this is clearer:

```suggestion
Parsers MUST fail parsing if values are not within the range supported by the parser.
```

I think that's simpler.

That is, say that parsers need to support up to 2^53-1.  Allow more.  Require parsing failure for numbers outside that range (whatever it is).

Author: Took a slightly different approach.

   6. If type is "float" and input_number contains more than 16 characters, fail parsing.
8. If type is "integer":

>    1. Parse input_number as an integer and let output_number be the product of the result and sign.

Reviewer 3: I totally agree with Kazuho here. It's mandatory to check for overflows during the parsing, like we do here for example : http://git.haproxy.org/?p=haproxy.git;a=blob;f=src/h1.c;hb=HEAD#l52
Many integer parsers will reliably provide the information about the fact that an overflow was detected, but when it comes to floats it might not necessarily be the case (since very large numbers can still fit in a float), however the check against bounds is expected to work.

This means that we should mandate the use of a reliable parser known for detecting overflows, AND then check for the output number to fall within our bounds.

We should also warn against the risk of using generic parsers which as a byproduct of checking for errors, also happen to support base prefixes like "0x", or float parsers supporting exponents like "1e32" or special values like "inf", "-inf", "nan", all of which defeat length checks.

>    1. Parse input_number as an integer and let output_number be the product of the result and sign.

Reviewer 4: > We should also warn against the risk of using generic parsers which as a byproduct of checking for errors, also happen to support base prefixes like "0x", or float parsers supporting exponents like "1e32" or special values like "inf", "-inf", "nan", all of which defeat length checks.

At this point in the algorithm the string has been preprocessed and is guaranteed to have no more than 19 characters, all of which are decimal digits. Or do you mean explicit warning against replacing the entire integer-parsing algorithm with a generic `to_int()` or equivalent?

>    2. If output_number is outside the range for integers defined in {{integer}}, fail parsing.

Reviewer 2: The two lines could be one, as follows:
```suggestion
   2. If output_number is outside the range for integers supported by the parser (see {{integer}}), fail parsing.
```

Author: I'd rather be explicit.

Reviewer 2: In which case, your two points need to be switched.  As written, this would prohibit supporting a larger range.

Reviewer 2: Belay that.  The comment stands.  There is no way to have a fail parsing on this condition.

Author: I don't think it prohibits supporting a larger range directly, but it does beg the question of how you can check that the number falls within the greater range if you can't accurately represent it. Will fix.

>    3. If output_number is outside the range for integers defined in {{integer}}, fail parsing.

Reviewer 2: As noted, this needs another condition ...and the implementation does not support integers outside this range.

Author: Where?

Reviewer 2: Let's say that I support -2^63 to 2^63-1, something that isn't unreasonable if you aren't javascript.  This code turns into:

```js
if (n < -2^63 || n > 2^63-1) { // Step 2.
  return fail_parsing;
}
if (n < -2^53+1 || n > 2^53-1) { // Step 3.
  return fail_parsing;
}
```

Pass n = 2^60 into that code and you get an error.  Hence my insistence that the first test is adequate.

Now, you could read "the range for integers" as meaning  -2^63 to 2^63-1 in this case, but then you just repeated yourself.

Does that help clarify the concern?
